modules/charon-core/src/test/java/org/wso2/charon3/core/utils/AttributeUtilTest.java

```diff
@@ -39,12 +50,145 @@
+
+        Instant instant = Instant.parse("2021-04-20T09:06:19.839Z");
+
+        return new Object[][]{
+
+                {"username", STRING, "username"},
+                {true, BOOLEAN, "true"},
+                {2, INTEGER, "2"},
+                {10.0, DECIMAL, "10.0"},
+                {instant, DATE_TIME, "2021-04-20T09:06:19.839Z"},
+                {0b010110, BINARY, "22"},
+                {"referenceString", REFERENCE, "referenceString"},
+                {"complexString", COMPLEX, "complexString"}
+        };
+    }
+
+    @Test(dataProvider = "dataForGetStringValueOfAttribute")
+    public void testGetStringValueOfAttribute(Object attributeValue, SCIMDefinitions.DataType
+            dataType, Object expectedStringValueOfAttribute) throws CharonException {
+
+        Object stringValueOfAttribute = AttributeUtil.getStringValueOfAttribute(attributeValue, dataType);
+        Assert.assertEquals(stringValueOfAttribute, expectedStringValueOfAttribute);
+    }
+
+    @DataProvider(name = "dataForParseDateTimeSuccess")
+    public Object[][] dataToParseDateTimeSuccess() {
+
+        Instant instant = Instant.parse("2021-04-20T09:06:19.839Z");
+
+        return new Object[][]{
+
+                {"2021-04-20T09:06:19.839Z", instant},
+                {"", null}
+        };
+    }
+
+    @Test(dataProvider = "dataForParseDateTimeSuccess")
+    public void testParseDateTimeSuccess(String dateTimeString, Instant expectedLocalDateTime)
+            throws CharonException {
+
+        Instant localDateTime = AttributeUtil.parseDateTime(dateTimeString);
+        Assert.assertEquals(localDateTime, expectedLocalDateTime);
+    }
+
+    @DataProvider(name = "dataForParseDateTimeExceptions")
+    public Object[][] dataToParseDateTimeExceptions() {
+
+        return new Object[][]{
+
+                {"2021-04-20T09"}
+        };
+    }
+
+    @Test(dataProvider = "dataForParseDateTimeExceptions")
+    public void testParseDateTimeExceptions(String dateTimeString) {
+
+        Assert.assertThrows(CharonException.class, () -> AttributeUtil.parseDateTime(dateTimeString));
+    }
+
+    @DataProvider(name = "dataForParseBooleanSuccess")
+    public Object[][] dataToParseBooleanSuccess() {
+
+        return new Object[][]{
+
+                {true, true},
+                {false, false},
+                {"true", true},
+                {"false", false},
+                {"abc", false},
+                {"", false},
+                {null, null}
+        };
+    }
+
+    @Test(dataProvider = "dataForParseBooleanSuccess")
+    public void testParseBooleanSuccess(Object booleanValueObject, Boolean expectedBooleanValue)
+            throws BadRequestException {
+
+        Boolean booleanValue = AttributeUtil.parseBoolean(booleanValueObject);
+        Assert.assertEquals(booleanValue, expectedBooleanValue);
+    }
+
+    @DataProvider(name = "dataForParseBooleanExceptions")
+    public Object[][] dataToParseBooleanExceptions() {
```

Comment:
Better if we can add a comment with what this data holder contains

Reply:
Fixed with b16f4bc341990ed28db680f0558364ff419804a4